Mathlib/Computability/AkraBazzi/GrowsPolynomially.lean

```diff
@@ -0,0 +1,340 @@
+/-
+Copyright (c) 2023 Frédéric Dupuis. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Frédéric Dupuis
+-/
+
+import Mathlib.Analysis.Asymptotics.AsymptoticEquivalent
+import Mathlib.Analysis.SpecialFunctions.Pow.Real
+
+/-!
+# Akra-Bazzi theorem: The polynomial growth condition
+
+This file defines and develops an API for the polynomial growth condition that appears in the
+statement of the Akra-Bazzi theorem: for the Akra-Bazzi theorem to hold, the function `g` must
+satisfy the condition that c₁ g(n) ≤ g(u) ≤ c₂ g(n), for u between b*n and n for any constant
+b ∈ (0,1).
+-/
+
+set_option autoImplicit false
+
+open Finset Real Filter Asymptotics BigOperators
+
+namespace AkraBazziRecurrence
+
+/-- The growth condition that the function `g` must satisfy for the Akra-Bazzi theorem to apply.
+It roughly states that c₁ g(n) ≤ g(u) ≤ c₂ g(n), for u between b*n and n for any
+constant b ∈ (0,1). -/
+def GrowsPolynomially (f : ℝ → ℝ) : Prop :=
+  ∀ b ∈ Set.Ioo 0 1, ∃ c₁ ∈ Set.Ioi 0, ∃ c₂ ∈ Set.Ioi 0,
+    ∀ᶠ x in atTop, ∀ u ∈ Set.Icc (b * x) x, f u ∈ Set.Icc (c₁ * (f x)) (c₂ * f x)
+
+namespace GrowsPolynomially
+
+variable {f : ℝ → ℝ} (hf : GrowsPolynomially f)
+
+lemma eventually_atTop_le {b : ℝ} (hb : b ∈ Set.Ioo 0 1) :
+    ∃ c ∈ Set.Ioi 0, ∀ᶠ x in atTop, ∀ u ∈ Set.Icc (b * x) x, f u ≤ c * f x := by
+  obtain ⟨c₁, _, c₂, hc₂, h⟩ := hf b hb
+  refine ⟨c₂, hc₂, ?_⟩
+  filter_upwards [h]
+  exact fun _ H u hu => (H u hu).2
+
+lemma eventually_atTop_le_nat {b : ℝ} (hb : b ∈ Set.Ioo 0 1) :
+    ∃ c ∈ Set.Ioi 0, ∀ᶠ (n:ℕ) in atTop, ∀ u ∈ Set.Icc (b * n) n, f u ≤ c * f n := by
+  obtain ⟨c, hc_mem, hc⟩ := hf.eventually_atTop_le hb
+  exact ⟨c, hc_mem, hc.nat_cast_atTop⟩
+
+lemma eventually_atTop_ge {b : ℝ} (hb : b ∈ Set.Ioo 0 1) :
+    ∃ c ∈ Set.Ioi 0, ∀ᶠ x in atTop, ∀ u ∈ Set.Icc (b * x) x, c * f x ≤ f u := by
+  obtain ⟨c₁, hc₁, c₂, _, h⟩ := hf b hb
+  refine ⟨c₁, hc₁, ?_⟩
+  filter_upwards [h]
+  exact fun _ H u hu => (H u hu).1
+
+lemma eventually_atTop_ge_nat {b : ℝ} (hb : b ∈ Set.Ioo 0 1) :
+    ∃ c ∈ Set.Ioi 0, ∀ᶠ (n:ℕ) in atTop, ∀ u ∈ Set.Icc (b * n) n, c * f n ≤ f u := by
+  obtain ⟨c, hc_mem, hc⟩ := hf.eventually_atTop_ge hb
+  exact ⟨c, hc_mem, hc.nat_cast_atTop⟩
+
+protected lemma const {c : ℝ} : GrowsPolynomially (fun _ => c) := by
+  intro b _
+  refine ⟨1, by norm_num, ?_⟩
+  refine ⟨1, by norm_num, ?_⟩
+  filter_upwards [] with x
+  intro u _
+  simp
+
+protected lemma mul {f g : ℝ → ℝ} (hf : GrowsPolynomially f) (hg : GrowsPolynomially g)
+    (hf' : 0 ≤ᶠ[atTop] f) (hg' : 0 ≤ᶠ[atTop] g) : GrowsPolynomially fun x => f x * g x := by
```

Comment:
Are the assumptions `hf'` and `hg'` really needed here? If a function grows polynomially, either it is eventually zero, or eventually positive, or eventually negative, and then everything goes through. Same thing in several lemmas below, all the eventual positivity assumptions can be dispensed with.

Reply:
Yes, that's probably right; I don't have time to fix this right now but I'll get back to it in the next few days hopefully.

Reply:
This is now fixed.